File: frontend/src/services/reservation.ts
import { get, post } from './api'
import type { ReservationListResponse } from '../types'

export function fetchReservations(date?: string): Promise<ReservationListResponse> {
  const qs = date ? `?date=${encodeURIComponent(date)}` : ''
  return get<ReservationListResponse>(`/reservations${qs}`)
}

export function createReservation(body: {
  user_name: string
  date: string
  start_hour: number
  end_hour: number
}) {
  return post<{ success: boolean; message: string; id: number }>('/reservations', body)
}

export function cancelReservation(id: number) {
  return post<{ success: boolean }>(`/reservations/${id}/cancel`)
}

File: backend/api/routes/reservation.py
"""
GET/POST /api/reservations — reservation management.
"""
from fastapi import APIRouter, Depends, HTTPException
from backend.api.deps import get_reservation_service
from backend.api.schemas import ReservationCreateRequest, ReservationListResponse

router = APIRouter(prefix="/api/reservations", tags=["reservations"])


@router.get("", response_model=ReservationListResponse)
def list_reservations(date: str = None, svc=Depends(get_reservation_service)):
    data = svc.get_reservations(date)
    return {"reservations": data}


@router.post("")
def create_reservation(body: ReservationCreateRequest, svc=Depends(get_reservation_service)):
    ok, msg, rid = svc.create_reservation(body.user_name, body.date, body.start_hour, body.end_hour)
    if not ok:
        raise HTTPException(status_code=409, detail=msg)
    return {"success": True, "message": msg, "id": rid}


@router.post("/{reservation_id}/cancel")
def cancel_reservation(reservation_id: int, svc=Depends(get_reservation_service)):
    svc.cancel_reservation(reservation_id)
    return {"success": True}
